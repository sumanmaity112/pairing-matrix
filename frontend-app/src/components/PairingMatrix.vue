<template>
  <ChordPairingChart
    :height="height"
    :width="width"
    :data="pairingMatrix"
    :authors="authors"
    v-if="displayChordChart"
  />
  <TabularPairingChart
    :height="height"
    :width="width"
    :data="pairingMatrix"
    :authors="authors"
    v-else
  />
</template>

<script>
import {
  ChordPairingChart,
  TabularPairingChart,
} from "pairing-matrix-vue-visualiser";

export default {
  name: "PairingMatrix",
  components: {
    ChordPairingChart,
    TabularPairingChart,
  },
  data() {
    return {
      pairingMatrix: [],
      authors: [],
      height: 954,
      width: 954,
    };
  },
  computed: {
    displayChordChart() {
      return this.chart === "chord";
    },
  },
  mounted() {
    this.fetchPairingMatrix();
  },
  methods: {
    fetchPairingMatrix() {
      return fetch(
        `./api/pair-matrix?since-days=${this.sinceDays}&pull-data=${this.pullData}&aggregate-by=${this.aggregateBy}`
      )
        .then((res) => res.json())
        .then(({ matrix, authors }) => {
          this.pairingMatrix = matrix;
          this.authors = authors;
        });
    },
  },
  props: {
    sinceDays: {
      type: Number,
      required: true,
    },
    pullData: {
      type: Boolean,
      required: true,
    },
    aggregateBy: {
      type: String,
      required: true,
    },
    chart: {
      type: String,
      required: true,
    },
  },
  watch: {
    sinceDays() {
      this.fetchPairingMatrix();
    },
    pullData() {
      this.fetchPairingMatrix();
    },
    aggregateBy() {
      this.fetchPairingMatrix();
    },
  },
};
</script>

<style scoped></style>
