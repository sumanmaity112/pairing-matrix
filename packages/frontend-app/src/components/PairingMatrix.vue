<template>
  <Spinner v-if="loading" />
  <Alert v-else-if="displayErrorMessage" :message="errorMessage" type="Error" />
  <Alert
    v-else-if="displayNoDataAvailable"
    :message="`No pairing data is available for last ${sinceDays} day(s)`"
    type="Info"
  />
  <div v-else>
    <div class="container">
      <PairingMatrixChart
        :height="height"
        :width="width"
        :pairingMatrix="pairingMatrix"
        :authors="authors"
        :chart="chart"
      />
      <PairRecommendation :recommendations="recommendations" />
    </div>
  </div>
</template>

<script>
import Spinner from "./Spinner";
import Alert from "./Alert";
import PairRecommendation from "@/components/PairRecommendation";
import PairingMatrixChart from "@/components/PairingMatrixChart";

export default {
  name: "PairingMatrix",
  components: {
    PairingMatrixChart,
    PairRecommendation,
    Spinner,
    Alert,
  },
  data() {
    return {
      pairingMatrix: [],
      authors: [],
      recommendations: {},
      height: 954,
      width: 954,
      loading: true,
      errorMessage: null,
    };
  },
  computed: {
    displayErrorMessage() {
      return !!this.errorMessage;
    },
    displayNoDataAvailable() {
      return this.authors.length === 0;
    },
  },
  mounted() {
    this.fetchPairingMatrix();
  },
  methods: {
    fetchPairingMatrix() {
      this.loading = true;
      this.errorMessage = null;
      return fetch(
        `./api/pair-matrix?since-days=${this.sinceDays}&pull-data=${this.pullData}&aggregate-by=${this.aggregateBy}`
      )
        .then((res) => (res.ok ? Promise.resolve(res) : Promise.reject(res)))
        .then((res) => res.json())
        .then(({ matrix, authors, recommendations }) => {
          this.loading = false;
          this.pairingMatrix = matrix;
          this.authors = authors;
          this.recommendations = recommendations;
        })
        .catch((res) => {
          this.loading = false;
          this.errorMessage = res.statusText;
        });
    },
  },
  props: {
    sinceDays: {
      type: Number,
      required: true,
    },
    pullData: {
      type: Boolean,
      required: true,
    },
    aggregateBy: {
      type: String,
      required: true,
    },
    chart: {
      type: String,
      required: true,
    },
  },
  watch: {
    sinceDays() {
      this.fetchPairingMatrix();
    },
    pullData() {
      this.fetchPairingMatrix();
    },
    aggregateBy() {
      this.fetchPairingMatrix();
    },
  },
};
</script>

<style scoped>
.container {
  display: flex;
}
</style>
