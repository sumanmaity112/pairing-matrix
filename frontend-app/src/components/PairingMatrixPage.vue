<template>
  <div class="pairing-matrix-page">
    <h2 class="pairing-matrix-title">Pairing Matrix</h2>
    <div class="pairing-matrix-container">
      <div class="pairing-matrix-config-container">
        <section class="pairing-matrix-config">
          <h4>Pull new data from github:</h4>
          <input type="radio" v-model="pullData" value="true" />Yes
          <input type="radio" v-model="pullData" value="false" />No
        </section>
        <section class="pairing-matrix-config">
          <h4>For last days:</h4>
          <input type="number" v-model="sinceDays" />
        </section>
        <section class="pairing-matrix-config">
          <h4>Aggregate By:</h4>
          <input type="radio" v-model="aggregateBy" value="issue" />Issue
          <input type="radio" v-model="aggregateBy" value="days" />Days
        </section>
        <section class="pairing-matrix-config">
          <h4>Visualise by:</h4>
          <input type="radio" v-model="chart" value="chord" />Chord Chart
          <input type="radio" v-model="chart" value="tabular" />Tabular Chart
        </section>
      </div>
      <div class="pairing-matrix">
        <PairingMatrix
          :aggregate-by="aggregateBy"
          :pull-data="shouldPullData"
          :since-days="sinceDays"
          :chart="chart"
        />
      </div>
    </div>
  </div>
</template>

<script>
import PairingMatrix from "@/components/PairingMatrix";

export default {
  name: "PairingMatrixPage",
  components: { PairingMatrix },
  data() {
    return {
      pullData: false,
      sinceDays: 14,
      aggregateBy: "issue",
      chart: "chord",
    };
  },
  computed: {
    shouldPullData() {
      return this.pullData === "true";
    },
  },
};
</script>

<style scoped>
.pairing-matrix-page {
  width: 80%;
  margin: 0 auto;
}

.pairing-matrix-title {
  text-align: center;
}

.pairing-matrix-container {
  display: flex;
}

.pairing-matrix-config-container {
  display: flex;
  flex-direction: column;
  width: 35%;
}

.pairing-matrix-config {
  margin-left: 20px;
  margin-right: 20px;
}

.pairing-matrix {
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 65%;
  margin: 0 auto;
}
</style>
