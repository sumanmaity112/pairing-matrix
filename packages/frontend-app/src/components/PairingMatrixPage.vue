<template>
  <div class="pairing-matrix-page">
    <h2 class="pairing-matrix-title">Pairing Matrix</h2>
    <div class="pairing-matrix-container">
      <div class="pairing-matrix-config-container">
        <section class="pairing-matrix-config">
          <h4>Pull new data from github:</h4>
          <input
            v-model="pullData"
            type="radio"
            value="true"
            data-testid="pairing-matrix-pull-data-config-yes"
          />Yes
          <input
            v-model="pullData"
            type="radio"
            value="false"
            data-testid="pairing-matrix-pull-data-config-no"
          />No
        </section>
        <section id="pairing-matrix-days-config" class="pairing-matrix-config">
          <h4>For last days:</h4>
          <input v-model="sinceDays" type="number" />
        </section>
        <section
          id="pairing-matrix-aggregation-config"
          class="pairing-matrix-config"
        >
          <h4>Aggregate By:</h4>
          <input v-model="aggregateBy" type="radio" value="issue" />Issue
          <input v-model="aggregateBy" type="radio" value="days" />Days
        </section>
        <section id="pairing-matrix-chart-config" class="pairing-matrix-config">
          <h4>Visualise by:</h4>
          <input v-model="chart" type="radio" value="chord" />Chord Chart
          <input v-model="chart" type="radio" value="tabular" />Tabular Chart
        </section>
      </div>
      <div class="pairing-matrix">
        <PairingMatrix
          :aggregate-by="aggregateBy"
          :pull-data="shouldPullData"
          :since-days="sinceDays"
          :chart="chart"
        />
      </div>
    </div>
  </div>
</template>

<script>
import PairingMatrix from "@/components/PairingMatrix";

export default {
  name: "PairingMatrixPage",
  components: { PairingMatrix },
  data() {
    return {
      // TODO: Change pullData to String
      pullData: true,
      sinceDays: 14,
      aggregateBy: "days",
      chart: "chord",
    };
  },
  computed: {
    shouldPullData() {
      return this.pullData === "true";
    },
  },
};
</script>

<style scoped>
.pairing-matrix-page {
  width: 93%;
  margin: 0 auto;
}

.pairing-matrix-title {
  text-align: center;
}

.pairing-matrix-container {
  display: flex;
}

.pairing-matrix-config-container {
  display: flex;
  flex-direction: column;
  width: 20%;
}

.pairing-matrix-config {
  margin-left: 20px;
  margin-right: 20px;
}

.pairing-matrix {
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 80%;
  margin: 0 auto;
}
</style>
