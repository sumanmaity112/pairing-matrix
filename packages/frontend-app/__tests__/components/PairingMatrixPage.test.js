import { shallowMount } from "@vue/test-utils";
import PairingMatrixPage from "@/components/PairingMatrixPage";
import PairingMatrix from "@/components/PairingMatrix";

describe("PairingMatrixPage", () => {
  const triggerRadioButtonClick = (radioInput) => {
    radioInput.trigger("click");
  };

  it("should render pairing matrix page with default configuration", () => {
    const wrapper = shallowMount(PairingMatrixPage);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it.each(["no"])(
    "should re-configure pullData configuration to %s",
    (suffix) => {
      const wrapper = shallowMount(PairingMatrixPage);

      triggerRadioButtonClick(
        wrapper.find(
          `[data-testid="pairing-matrix-pull-data-config-${suffix}"]`
        )
      );

      expect(wrapper.html()).toMatchSnapshot();
    }
  );
});
