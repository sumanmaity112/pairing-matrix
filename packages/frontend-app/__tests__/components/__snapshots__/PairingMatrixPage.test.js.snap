// Vitest Snapshot v1

exports[`PairingMatrixPage > should re-configure pullData configuration to no 1`] = `
<div class="pairing-matrix-page">
  <h2 class="pairing-matrix-title">Pairing Matrix</h2>
  <div class="pairing-matrix-container">
    <div class="pairing-matrix-config-container">
      <section class="pairing-matrix-config">
        <h4>Pull new data from github:</h4>
        <input type="radio" value="true">Yes
        <input type="radio" value="false">No
      </section>
      <section class="pairing-matrix-config" id="pairing-matrix-days-config">
        <h4>For last days:</h4>
        <input type="number">
      </section>
      <section class="pairing-matrix-config" id="pairing-matrix-aggregation-config">
        <h4>Aggregate By:</h4>
        <input type="radio" value="issue">Issue
        <input type="radio" value="days">Days
      </section>
      <section class="pairing-matrix-config" id="pairing-matrix-chart-config">
        <h4>Visualise by:</h4>
        <input type="radio" value="chord">Chord Chart
        <input type="radio" value="tabular">Tabular Chart
      </section>
    </div>
    <div class="pairing-matrix">
      <pairing-matrix-stub aggregateby="days" chart="chord" pulldata="false" sincedays="14"></pairing-matrix-stub>
    </div>
  </div>
</div>
`;

exports[`PairingMatrixPage > should render pairing matrix page with default configuration 1`] = `
<div class="pairing-matrix-page">
  <h2 class="pairing-matrix-title">Pairing Matrix</h2>
  <div class="pairing-matrix-container">
    <div class="pairing-matrix-config-container">
      <section class="pairing-matrix-config">
        <h4>Pull new data from github:</h4>
        <input type="radio" value="true">Yes
        <input type="radio" value="false">No
      </section>
      <section class="pairing-matrix-config" id="pairing-matrix-days-config">
        <h4>For last days:</h4>
        <input type="number">
      </section>
      <section class="pairing-matrix-config" id="pairing-matrix-aggregation-config">
        <h4>Aggregate By:</h4>
        <input type="radio" value="issue">Issue
        <input type="radio" value="days">Days
      </section>
      <section class="pairing-matrix-config" id="pairing-matrix-chart-config">
        <h4>Visualise by:</h4>
        <input type="radio" value="chord">Chord Chart
        <input type="radio" value="tabular">Tabular Chart
      </section>
    </div>
    <div class="pairing-matrix">
      <pairing-matrix-stub aggregateby="days" chart="chord" pulldata="false" sincedays="14"></pairing-matrix-stub>
    </div>
  </div>
</div>
`;
